lib/utility/src/Logger.cpp

```diff
@@ -1,143 +1,66 @@
 #include "Logger.h"
 
-namespace Utility {
-static char buffer[4096];  // Can increase max buffer length if needed
-}
-
-int Utility::Logger::printf(const char* format, ...) {
-  va_list arg;
-
-  va_start(arg, format);
-  int num_chars = std::vsprintf(buffer, format, arg);
-  va_end(arg);
-
-#if defined(UART_LOGGING)
-  std::printf(buffer);
-#endif
-
-#if defined(SWO_LOGGING)
-  char* c = buffer;
-
-  while (*c) {
-    ITM_SendChar(*c++);
-    num_chars++;
-  }
-#endif
-  return num_chars;
-}
-
-Utility::Logger& Utility::Logger::operator<<(long n) {
-  return operator<<(static_cast<long long>(n));
-}
-
-Utility::Logger& Utility::Logger::operator<<(unsigned long n) {
-  return operator<<(static_cast<unsigned long long>(n));
-}
-
-Utility::Logger& Utility::Logger::operator<<(bool n) {
-  return operator<<(static_cast<long long>(n));
-}
-
-Utility::Logger& Utility::Logger::operator<<(short n) {
-  return operator<<(static_cast<long long>(n));
-}
+#include "hal/itm_api.h"
+#include "hal/serial_api.h"
 
-Utility::Logger& Utility::Logger::operator<<(unsigned short n) {
-  return operator<<(static_cast<unsigned long long>(n));
-}
-
-Utility::Logger& Utility::Logger::operator<<(int n) {
-  return operator<<(static_cast<long long>(n));
-}
+extern serial_t stdio_uart;
 
-Utility::Logger& Utility::Logger::operator<<(unsigned int n) {
-  return operator<<(static_cast<unsigned long long>(n));
+namespace mbed {
+// Redirect stdin, stdout, and stderr to Logger
+FileHandle *mbed_target_override_console(int) {
+  static Utility::Logger logger;
+  return &logger;
 }
+}  // namespace mbed
 
-Utility::Logger& Utility::Logger::operator<<(long long n) {
-#if defined(UART_LOGGING)
-  std::printf("%lld", n);
-#endif
+namespace Utility {
 
-#if defined(SWO_LOGGING)
-  std::sprintf(buffer, "%lld", n);
-  char* c = buffer;
+Logger logger;
+Mutex Logger::m_mutex;
 
-  while (*c) {
-    ITM_SendChar(*c++);
+Logger::Logger() {
+  if constexpr (FeatureFlags::UART_LOGGING) {
```

Comment:
fyi: need to be very careful if we're gonna use this way of doing feature flags because it does emulate the normal `#if` and `#ifdef` way. The compiler will still process all the output in these blocks when they're false so I think it may cause us future grief with compiler warning and (if/when implemented) linter warnings.

From ccpreference:
![image](https://user-images.githubusercontent.com/16629657/110252050-87a8d780-7f51-11eb-842f-238344826e2f.png)
![image](https://user-images.githubusercontent.com/16629657/110252060-9099a900-7f51-11eb-9381-2148b7709275.png)

Reply:
Had a typo in my previous comment: it **doesnt** emulate preprocessor directive exactly